### PR TITLE
fix(react-mcp): validate oauth callback parameters

### PR DESCRIPTION
**Pull request by Kinfe123**
## Summary

- persist the generated OAuth state alongside its PKCE verifier
- reject callback state that does not match the active authorization request
- forward the complete callback parameters to the MCP SDK for issuer and error validation
- consume the one-time state after tokens are stored

## Why

React MCP currently uses the server ID embedded in `state` to route a callback, but it does not compare the complete state value with the request that initiated authorization. It also passes only the authorization code to `finishAuth()`, which prevents the MCP SDK from validating the callback issuer.

Persisting the exact state binds the callback to its authorization request. Passing `URLSearchParams` preserves `iss` and lets the SDK validate the issuer before redeeming a code or exposing callback error text.

## Verification

- `pnpm --filter @assistant-ui/react-mcp test` (12 files, 157 tests)
- `pnpm turbo build --filter=@assistant-ui/react-mcp...`
- focused Oxlint and Oxfmt checks
- API surface generation and type check
- `git diff --check`

### CODE REVIEW

File: packages/react-mcp/src/auth/createOAuthProvider.ts [hunk 1]
```diff
@@ -246,6 +252,10 @@ export function createOAuthProvider(
     async saveCodeVerifier(codeVerifier) {
       const c = await loadCache();
       c.codeVerifier = codeVerifier;
+      if (pendingState) {
+        c.state = pendingState;
+        pendingState = undefined;
+      }
       await persist();
     },
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-mcp/src/auth/createOAuthProvider.ts:260]: Addressed in 405b41b5e. The existing real SDK auth test now asserts that the state placed on the authorization URL is persisted with the PKCE verifier. I also updated the storage and OAuth flow documentation for the new persisted field.

File: packages/react-mcp/src/resources/McpServerResource.ts [hunk 2]
```diff
@@ -476,8 +476,15 @@ const useMcpServerResourceInstance = (
...
       const url = new URL(callbackUrl);
-      const code = url.searchParams.get("code");
-      if (!code) throw new Error("missing authorization code in callback URL");
+      const state = url.searchParams.get("state");
+      if (!state) throw new Error('missing "state" parameter');
+      const persisted = await props.storage.loadAuthState(props.id);
+      if (!persisted?.state || persisted.state !== state) {
+        throw new Error("OAuth state does not match the authorization request");
+      }
+      if (!url.searchParams.has("code") && !url.searchParams.has("error")) {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-mcp/src/resources/McpServerResource.ts]: Fixed in fc19a23b7. Empty code and error values are now treated as missing before transport setup, and the regression test covers an empty code parameter.

File: packages/react-mcp/src/resources/McpServerResource.ts [hunk 3]
```diff
@@ -476,8 +476,15 @@ const useMcpServerResourceInstance = (
...
     try {
       const url = new URL(callbackUrl);
-      const code = url.searchParams.get("code");
-      if (!code) throw new Error("missing authorization code in callback URL");
+      const state = url.searchParams.get("state");
+      if (!state) throw new Error('missing "state" parameter');
+      const persisted = await props.storage.loadAuthState(props.id);
+      if (!persisted?.state || persisted.state !== state) {
+        throw new Error("OAuth state does not match the authorization request");
+      }
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-mcp/src/resources/McpServerResource.ts]: Addressed in 35d011bdb. Missing persisted state now reports that no pending authorization request exists, while a different persisted value retains the mismatch error. A focused test covers the missing-state path.

File: packages/react-mcp/SPEC.md [hunk 4]
```diff
@@ -267,14 +268,14 @@ type MCPAuthConfig =
...
-The server id is embedded in the OAuth `state` parameter so a single `/mcp/callback` route routes back to the right server without app-level wiring.
+The server id is embedded in the OAuth `state` parameter so a single `/mcp/callback` route routes back to the right server without app-level wiring. The complete state value is persisted with the PKCE verifier and validated before the callback is processed.
 
 Flow:
 
 1. `aui.mcp().server({ id }).connect()` → SDK starts auth.
 2. `redirectToAuthorization` stores `authorizationUrl` on server state, transitions to `authRequired`. **The package does not auto-navigate** — render `<McpServerPrimitive.OAuthLink>` (an anchor) or open a popup.
 3. User returns to `oauthRedirectUri`. Mount `<McpOAuthCallback />` there.
-4. Callback reads `?state=&code=`, derives the server id, calls `server.completeAuth(window.location.href)`.
+4. Callback reads `?state=&code=`, derives the server id, validates the complete state value, and calls `server.completeAuth(window.location.href)`.
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-mcp/SPEC.md]: Fixed in 781d453f1. The flow now documents both successful code callbacks and OAuth error callbacks, matching completeAuth and the lifecycle section.

File: packages/react-mcp/src/resources/McpServerResource.ts [hunk 5]
```diff
@@ -476,8 +476,20 @@ const useMcpServerResourceInstance = (
     setLastError(null);
     try {
       const url = new URL(callbackUrl);
-      const code = url.searchParams.get("code");
-      if (!code) throw new Error("missing authorization code in callback URL");
+      const state = url.searchParams.get("state");
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-mcp/src/resources/McpServerResource.ts]: Addressed in e728f79b8. Callback parsing, persisted-state comparison, and code/error validation now happen before the connection generation changes or any transport/elicitation teardown. The regression test starts with a connected server and verifies a forged state rejects without closing its transport or changing its state.

File: packages/react-mcp/src/resources/McpServerResource.ts [hunk 6]
```diff
@@ -467,6 +467,20 @@ const useMcpServerResourceInstance = (
   });
 
   const doCompleteAuth = useEffectEvent(async (callbackUrl: string) => {
+    const url = new URL(callbackUrl);
+    const state = url.searchParams.get("state");
+    if (!state) throw new Error('missing "state" parameter');
+    const persisted = await props.storage.loadAuthState(props.id);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-mcp/src/resources/McpServerResource.ts:474]: Fixed in 5e5b7e6f1. completeAuth now captures the connection generation before loading persisted state and rejects as interrupted if disconnect or another lifecycle change advances it. The regression test disconnects while loadAuthState is pending and verifies no transport is created and the server remains disconnected.

File: packages/react-mcp/src/resources/McpServerResource.test.ts [hunk 7]
```diff
@@ -575,26 +579,106 @@ describe("McpServerResource completeAuth", () => {
...
+      expect(params.get("iss")).toBe("https://auth.example.com");
       expect(root.getValue().getState().connectionState).toBe("connected");
     } finally {
       root.unmount();
     }
   });
 
-  it("rejects when the callback URL has no authorization code", async () => {
-    const root = mount();
+  it("ignores callback URLs without a usable authorization code", async () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-mcp/src/resources/McpServerResource.test.ts]: Fixed in 5e5b7e6f1. The test now says it rejects before transport setup, matching both the assertion and the behavior under test.

File: apps/docs/content/docs/tools/user-managed-mcp.mdx [hunk 8]
```diff
@@ -297,7 +297,7 @@ Client-side validation errors keep the elicitation pending so the user can corre
 
 ## Storage
 
-All persisted state — custom server records, OAuth tokens, PKCE verifiers, DCR client info — goes through a single `MCPStorage` resource. Three built-ins:
+All persisted state — custom server records, OAuth tokens, callback state, PKCE verifiers, DCR client info — goes through a single `MCPStorage` resource. Three built-ins:
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply apps/docs/content/docs/tools/user-managed-mcp.mdx:300]: Fixed in 5e5b7e6f1. The user-facing OAuth flow now documents code and error callbacks and the exact pending-state validation, matching the implementation and SPEC.

**Comment by okisdev** [comment]
heads-up now that #6447 has landed: this needs a rebase, and my note over there that it would merge cleanly was half wrong.

`useMcpOAuthCallback.tsx` does auto-merge, since #6447 only changed the `optsRef` publication above the async body and this removes the inline `error` parameter handling inside it. `useMcpOAuthCallback.test.ts` conflicts, because both PRs appended cases to the same region.

resolving the test file is a keep-both: #6447 added `keeps callbacks scoped to committed renders`, which drives a real interrupted render through Suspense and needs its `createElement`/`Suspense`/`startTransition`/`act` imports preserved alongside whatever this PR adds.

File: packages/react-mcp/src/resources/McpServerResource.ts [hunk 9]
```diff
@@ -467,6 +467,24 @@ const useMcpServerResourceInstance = (
   });
 
   const doCompleteAuth = useEffectEvent(async (callbackUrl: string) => {
+    const validationGeneration = connectionGenerationRef.current;
+    const url = new URL(callbackUrl);
+    const state = url.searchParams.get("state");
+    if (!state) throw new Error('missing "state" parameter');
+    const persisted = await props.storage.loadAuthState(props.id);
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/react-mcp/src/resources/McpServerResource.ts:474]: confirmed against main: the bump really was the first statement of doCompleteAuth, so the callback did fence auto-connect out, and hoisting the validation gives that up.

taking it as a follow-up rather than holding this PR. the preconditions are narrow (persisted tokens on the callback page plus a storage read slow enough to lose the order), and the fix is a new in-flight ref threaded through tryAutoConnect, which is a different concern from validating callback parameters.

tracked in #6535.

**Review by okisdev — APPROVED** [review]
approving. the premise here is upstream-endorsed rather than defensive hardening, which is what settles it: the jsdoc on `finishAuth` in `@modelcontextprotocol/client@2.0.0` says outright that the SDK does not validate `state` and that the client has to compare it before calling, and `resolveAuthorizationCallbackParams` validates `iss` before it reads any of the callback's error text. both halves of this PR close a gap the SDK deliberately leaves to us.

the part I checked hardest is the persistence binding, since stashing the generated state in `state()` and only writing it in `saveCodeVerifier()` is safe only if the SDK always calls them in that order. it does: `auth()` runs `provider.state()`, then `startAuthorization()`, then `saveCodeVerifier()`, then `redirectToAuthorization()`, so the state always lands in the same storage write as the verifier it belongs to and never outlives a flow that never started. that is a read of the installed dist, not something I exercised.

I pushed two commits. the first is a test defect rather than polish: `forwards OAuth error callbacks to the transport` let the mocked `finishAuth` resolve, so it walked on through `finalizeConnect` and asserted a server sitting in `connected` after a denied authorization, which the shipped path cannot reach because `resolveAuthorizationCallbackParams` has no non-throwing exit when `code` is absent. it now rejects the way the `invalid_grant` test does and pins `connectionState: "error"`. the second adds a clause to SPEC.md §6, because the pre-transport checks deliberately reject without touching `connectionState` or `lastError` while that section already documents the opposite path for a rejected `loadAuthState` during auto-connect, and two paragraphs reading differently is how someone later "fixes" the silence.

`pnpm --filter @assistant-ui/react-mcp test` on the head prints `Test Files 12 passed (12)` and `Tests 159 passed (159)`.

one thing stays open, tracked in #6535 rather than held against this PR. hoisting the validation above the `++connectionGenerationRef` is the right call, but it gives up a fence: the bump used to be the first statement of `doCompleteAuth`, so a callback always won the mount race against `tryAutoConnect`. both sides are now pending storage reads, and if auto-connect's resolves first while tokens are already persisted it calls `doConnect()`, bumps the generation, and the callback rejects as interrupted without ever reaching `finishAuth`. it needs persisted tokens plus a storage read slow enough to lose the order, so `McpLocalStorage` will rarely show it and a network-backed `McpCustomStorage` genuinely can. restoring the synchronous bump is not the fix, since that is what reintroduces the forged-callback teardown; a separate in-flight ref is.
